apps/webapp/app/v3/marqs.server.ts

```diff
@@ -0,0 +1,906 @@
+  async #calculateQueueWeights(
+    queues: Array<{ value: string; score: number }>,
+    calculateWeight: (queue: string, score: number) => Promise<number>
+  ) {
+    const queueWeights = await Promise.all(
+      queues.map(async (queue) => {
+        return {
+          queue: queue.value,
+          weight: await calculateWeight(queue.value, queue.score),
+        };
+      })
+    );
+
+    return queueWeights;
+  }
+
+  async #calculateMessageQueueWeight(queue: string, score: number) {
+    const concurrencyQueueName = queue.replace(/:ck:.+$/, "");
+
+    const concurrencyLimit =
+      (await this.redis.get(`${concurrencyQueueName}:${constants.CONCURRENCY_LIMIT_PART}`)) ?? 100;
+    const currentConcurrency = await this.redis.scard(
+      `${queue}:${constants.CURRENT_CONCURRENCY_PART}`
+    );
+
+    const capacity = Number(concurrencyLimit) - Number(currentConcurrency);
+
+    const capacityWeight = capacity / Number(concurrencyLimit);
+    const ageWeight = Date.now() - score;
+
+    return ageWeight * 0.8 + capacityWeight * 0.2;
+  }
+
+  async #weightedRandomChoice(queues: Array<{ queue: string; weight: number }>) {
+    const totalWeight = queues.reduce((acc, queue) => acc + queue.weight, 0);
+    const randomNum = Math.random() * totalWeight;
+    let weightSum = 0;
+
+    for (const queue of queues) {
+      weightSum += queue.weight;
+      if (randomNum <= weightSum) {
+        return queue.queue;
+      }
+    }
+
+    return queues[queues.length - 1].queue;
+  }
+
+  async #zrangeWithScores(
+    key: string,
+    min: number,
+    max: number
+  ): Promise<Array<{ value: string; score: number }>> {
+    const valuesWithScores = await this.redis.zrange(key, min, max, "WITHSCORES");
+    const result: Array<{ value: string; score: number }> = [];
+
+    for (let i = 0; i < valuesWithScores.length; i += 2) {
+      result.push({
+        value: valuesWithScores[i],
+        score: Number(valuesWithScores[i + 1]),
+      });
+    }
+
+    return result;
+  }
+
+  #startRequeuingWorkers() {
+    // Start a new worker to requeue visible messages
+    for (let i = 0; i < this.options.workers; i++) {
+      const worker = new AsyncWorker(this.#requeueVisibleMessages.bind(this), 1000);
+
+      this.#requeueingWorkers.push(worker);
+
+      worker.start();
+    }
+  }
+
+  async #requeueVisibleMessages() {
+    // Remove any of the messages from the timeoutQueue that have expired
+    const messages = await this.redis.zrangebyscore(
+      constants.MESSAGE_VISIBILITY_TIMEOUT_QUEUE,
+      0,
+      Date.now(),
+      "LIMIT",
+      0,
+      10
+    );
+
+    if (messages.length === 0) {
+      return;
+    }
+
+    for (let i = 0; i < messages.length; i++) {
+      const message = messages[i];
+
+      const messageData = await this.redis.get(`${constants.MESSAGE_PART}:${message}`);
+
+      if (!messageData) {
+        await this.redis.zrem(constants.MESSAGE_VISIBILITY_TIMEOUT_QUEUE, message);
```

Comment:
If there's no message data what does that mean?

If it's a bug then we should log

Reply:
A code comment would be useful if it's not a bug